azurerm/resource_arm_monitor_diagnostics.go

```diff
@@ -0,0 +1,284 @@
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+	log.Printf("[INFO] preparing arguments for Azure ARM Diagnostic Settings.")
+
+	name := d.Get("name").(string)
+	targetResourceId := d.Get("target_resource_id").(string)
+	storageAccountId := d.Get("storage_account_id").(string)
+	eventHubName := d.Get("event_hub_name").(string)
+	eventHubAuthorizationRuleId := d.Get("event_hub_authorization_rule_id").(string)
+	workspaceId := d.Get("workspace_id").(string)
+	disabledSettings := d.Get("disabled_settings").([]interface{})
+	retentionDays := d.Get("retention_days").(int)
+
+	allMetricSettings, allLogSettings, err := getAllDiagnosticSettings(targetResourceId, meta)
+	if err != nil {
+		return err
+	}
+
+	allSettings := append(*allMetricSettings, *allLogSettings...)
+	if !utils.StringSliceContainsStringSlice(disabledSettings, allSettings) {
+		return fmt.Errorf("Invalid value for disabled settings provided, use one or multiple of: %q", allSettings)
+	}
+
+	if len(allSettings) == len(disabledSettings) {
+		return fmt.Errorf("You can not disable all settings, rather delete diagnostic logging")
+	}
+
+	diagnosticSettings := &insights.DiagnosticSettings{}
+	diagnosticSettings.Metrics = expandMetricsConfiguration(*allMetricSettings, disabledSettings, retentionDays)
+	diagnosticSettings.Logs = expandLogConfiguration(*allLogSettings, disabledSettings, retentionDays)
+
+	if len(storageAccountId) > 0 {
+		diagnosticSettings.StorageAccountID = &storageAccountId
+	}
+
+	if len(workspaceId) > 0 {
+		diagnosticSettings.WorkspaceID = &workspaceId
+	}
+
+	if len(eventHubAuthorizationRuleId) > 0 && len(eventHubName) > 0 {
+		diagnosticSettings.EventHubAuthorizationRuleID = &eventHubAuthorizationRuleId
+		diagnosticSettings.EventHubName = &eventHubName
+	}
+
+	_, err = client.CreateOrUpdate(
+		ctx,
+		targetResourceId,
+		insights.DiagnosticSettingsResource{
+			DiagnosticSettings: diagnosticSettings,
+		},
+		name)
+	if err != nil {
+		return fmt.Errorf("Error creating Diagnostics Setting %q (Resource ID %q): %+v", name, targetResourceId, err)
+	}
+
+	read, err := client.Get(ctx, targetResourceId, name)
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read ID for Monitor Diagnostics %q", name)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmMonitorDiagnosticsRead(d, meta)
+}
+
+func resourceArmMonitorDiagnosticsRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).monitorDiagnosticSettingsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	monitoringId := parseMonitorDiagnosticId(d.Id())
+
+	resp, err := client.Get(ctx, monitoringId.ResourceID, monitoringId.Name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on Diagnostic Setting %s: %+v", monitoringId.Name, err)
+	}
+
+	d.Set("name", *resp.Name)
```

Comment:
FYI `d.Set` will take care of dereferencing the pointer as needed; so we can remove the `*` here (I'll push a commit with this)